rust/kernel/cpu.rs

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! CPU utilities for SMP systems.
+//!
+//! C header: [`include/linux/smp.h`](../../../../include/linux/smp.h).
+//!
+//! Note: This module is a work in progress.
+
+use crate::{bindings, error::code::*, prelude::*, to_result};
+use core::{marker::PhantomData, ops::Deref};
+
+/// Represents an SMP CPU by its zero-based kernel processor id.
+///
+/// Cpu values can be safely passed between threads (and across physical CPUs) as they are
+/// merely an integer index representing a CPU. All actual interaction with CPUs is late-
+/// bound.
+#[derive(Copy, Clone)]
+pub struct Cpu(u32);
+
+struct FnCall<F, R>
+where
+    F: FnOnce() -> R,
+{
+    out: Option<R>,
+    func: Option<F>,
+}
+
+unsafe extern "C" fn call_fn<F, R>(fn_box: *mut core::ffi::c_void)
+where
+    F: FnOnce() -> R,
+{
+    // SAFETY: The casting here is the inverse of the cast to *mut c_void that occurs in Cpu::call
+    // below. That call waits for this call to complete ensuring fn_call is live for the duration.
```

Comment:
I think @wedsonaf mentioned this above for some other bit of code, but we use Markdown in comments too. This makes things more consistent and allows easier copy-pasting between docs and comments:

```suggestion
    // SAFETY: The casting here is the inverse of the cast to `*mut c_void` that occurs in `Cpu::call`
    // below. That call waits for this call to complete ensuring `fn_call` is live for the duration.
```